File: main/ODE_Module_main.cpp

#include <iostream>
#include <vector>
#include <string>
#include <filesystem>

#include "../include/ODE_Module/ODE_Module.hpp"
#include "../include/Utilities.hpp"

using namespace std::filesystem;
using namespace ScientificToolbox::ODE;

int main(int argc, char** argv) {
    try {
        // Default values
        path project_dir = std::filesystem::current_path();
        path inputFile = project_dir / path("data") / path("ode_examples.csv");
        path outputFolder = project_dir / path("output");

        std::cout << "Enter input filename from data folder (press Enter for default):\n";
        std::string userInput;
        std::getline(std::cin, userInput);
        if (!userInput.empty()) {
            inputFile = project_dir / "data" / userInput;
        }

        std::cout << "ODE Module Demo\n" << std::endl;

        // You can load the data from a CSV file using the Importer class
        std::vector<ODETestCase> test_cases = load_tests_from_csv(inputFile);

        /*
         * NOTE: 
         *   The ODE solvers return an ODESolution object:
         *       you can modify the number of steps to print changing
         *       the value of ODESolution.steps_to_print [default = 10]) 
         *       Use ODESolution.steps_to_print = ODESolution.size to print all steps
         *       Use the get_result() method to access the final value
         */

        std::cout << std::endl;
        for (const auto& test : test_cases) {
            std::cout << std::endl << test << std::endl;
            double t0 = test.t0;
            double tf = test.tf;
            double h = test.h;
            var_vec y0 = test.y0; 

            std::cout << std::endl << "  Solving ODE using different methods:" << std::endl << std::endl;
            
            // reset files
            std::ofstream(outputFolder / "ODe_Module" / "euler_solution.csv", std::ios::trunc);
            std::ofstream(outputFolder / "ODe_Module" / "midpoint_solution.csv", std::ios::trunc);
            std::ofstream(outputFolder / "ODe_Module" / "rk4_solution.csv", std::ios::trunc);

            // Solve using Forward Euler
            auto euler = ForwardEulerSolver(test.expr, y0, t0, tf, h);
            auto euler_solution = euler.solve();
            std::cout << "  Forward Euler Method:\t\t";
            // use the result field to access the final value
            std::cout << euler_solution.get_result() << std::endl;
            save_to_csv(outputFolder / "euler_solution.csv", euler_solution, true);

            // Solve using Explicit Midpoint
            auto midpoint = ExplicitMidpointSolver(test.expr, y0, t0, tf, h);
            auto midpoint_solution = midpoint.solve();
            std::cout << "  Explicit Midpoint Method:\t";
            // use the result field to access the final value
            std::cout << midpoint_solution.get_result() << std::endl;
            save_to_csv(outputFolder / "midpoint_solution.csv", midpoint_solution, true);

            // Solve using RK4
            auto rk4 = RK4Solver(test.expr, y0, t0, tf, h);
            auto rk4_solution = rk4.solve();
            std::cout << "  Runge-Kutta 4th Order Method:\t";
            // use the result field to access the final value
            std::cout << rk4_solution.get_result() << std::endl;
            save_to_csv(outputFolder / "rk4_solution.csv", rk4_solution, true);
        }

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: main/ODE_Module_main.py
import os
import sys
import scientific_toolbox

# make shure you have the scientific_toolbox package installed with `pip install -e .`
try:
    import scientific_toolbox.ode as ode
except ImportError as e:
    print("\nModule not found. Have you built the project?\nRun 'make' in the build directory, then 'pip install -e .'\n")
    print(e)
    sys.exit(1)

ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
data_dir = os.path.join(ROOT_DIR, 'data')
output_dir = os.path.join(ROOT_DIR, 'output/ODE_Module/Analysis')

def main():
    # Letting the user choose the path to data containing the ODEs, and if does not enter anything, the default data is used
    data_path = input("Enter the path to the csv file containing the ODEs: ")
    if data_path == "":
        data_path = os.path.join(data_dir, 'ode_tests.csv')
        print(f"No path was entered, using the default ODEs in {data_path}")
        
    data = ode.load_tests_from_csv(data_path)
    
    os.makedirs(output_dir, exist_ok=True)
    
    output_files = {
        "euler"     : os.path.join(output_dir, "euler_solution.csv"),
        "midpoint"  : os.path.join(output_dir, "midpoint_solution.csv"),
        "rk4"       : os.path.join(output_dir, "rk4_solution.csv")
    }
    
    plot_folders = {
        "euler"     : os.path.join(output_dir, "euler_plots"),
        "midpoint"  : os.path.join(output_dir, "midpoint_plots"),
        "rk4"       : os.path.join(output_dir, "rk4_plots")
    }
    
    analyzer = ode.ODEAnalysis()
    
    # Clear the output files
    for file in output_files.values():
        with open(file, "w") as f:
            f.write("")
    
    for i, case in enumerate(data):
        print(case)
        
        euler = ode.ForwardEulerSolver(case)
        euler_solution = euler.solve()
        print(f"The euler solution of the ODE is:\t{euler_solution.get_result()}")
        ode.save_to_csv(output_files["euler"], euler_solution, append = True)
        analyzer.plot_solution(euler_solution, title = f"{case.get_expr()} - Euler", save_path = os.path.join(plot_folders["euler"], f"case_{i}"), show = False)
        
        midpoint = ode.ExplicitMidpointSolver(case)
        midpoint_solution = midpoint.solve()
        print(f"The midpoint solution of the ODE is:\t{midpoint_solution.get_result()}")
        ode.save_to_csv(output_files["midpoint"], midpoint_solution, append = True)
        analyzer.plot_solution(midpoint_solution, title = f"{case.get_expr()} - Midpoint", save_path = os.path.join(plot_folders["midpoint"], f"case_{i}"), show = False)
        
        rk4 = ode.RK4Solver(case)
        rk4_solution = rk4.solve()
        print(f"The RK4 solution of the ODE is:\t\t{rk4_solution.get_result()}")       
        ode.save_to_csv(output_files["rk4"], rk4_solution, append = True)
        analyzer.plot_solution(rk4_solution, title = f"{case.get_expr()} - RK4", save_path = os.path.join(plot_folders["rk4"], f"case_{i}"), show = False)
        
        print()
    
if __name__ == '__main__':
    main()